netadapter/server/grpcserver/connection_loops.go

```diff
@@ -29,39 +28,45 @@ func (c *gRPCConnection) connectionLoops(stream grpcStream) error {
 
 func (c *gRPCConnection) sendLoop(stream grpcStream) error {
 	for c.IsConnected() {
-		message, ok := <-c.sendChan
-		if !ok {
-			return nil // this means the sendChan is closed, a.k.a. connection is disconnecting
+		message, err := c.router.OutgoingRoute().Dequeue()
```

Comment:
I'd do `outgoingRoute := c.router.OutgoingRoute()` outside the loop
And then here do `message, err := outgoingRoute.Dequeue()`